datacube/api/grid_workflow.py

```diff
@@ -0,0 +1,459 @@
+# This file is part of the Open Data Cube, see https://opendatacube.org for more information
+#
+# Copyright (c) 2015-2025 ODC Contributors
+# SPDX-License-Identifier: Apache-2.0
+from __future__ import annotations
+from collections.abc import Hashable
+import logging
+import typing
+import xarray as xr
+from collections import OrderedDict
+import pandas as pd
+
+from odc.geo.geom import intersects
+from .query import Query, query_group_by
+from .core import Datacube
+
+_LOG = logging.getLogger(__name__)
+
+if typing.TYPE_CHECKING:
+    from odc.geo.geobox import GeoBox
+    from datacube.model import Product
+    from datacube.index import Index
+
+
+def _fast_slice(array: xr.DataArray, indexers):
+    data = array.values[indexers]
+    dims = [
+        dim for dim, indexer in zip(array.dims, indexers) if isinstance(indexer, slice)
+    ]
+    coords = OrderedDict(
+        (
+            dim,
+            xr.Variable(
+                (dim,),
+                array.coords[dim].values[indexer],
+                attrs=array.coords[dim].attrs,
+                fastpath=True,
+            ),
+        )
+        for dim, indexer in zip(array.dims, indexers)
+        if isinstance(indexer, slice)
+    )
+    return xr.DataArray(data, dims=dims, coords=coords, attrs=array.attrs)
+
+
+class Tile:
+    """
+    The Tile object holds a lightweight representation of a datacube result.
+
+    It is produced by :meth:`.GridWorkflow.list_cells` or :meth:`.GridWorkflow.list_tiles`.
+
+    The Tile object can be passed to :meth:`GridWorkflow.load` to be loaded into memory as
+    an :class:`xr.Dataset`.
+
+    A portion of a tile can be created by using index notation. eg:
+
+        tile[0:1, 0:1000, 0:1000]
+
+    This can be used to load small portions of data into memory, instead of having to access
+    the entire `Tile` at once.
+    """
+
+    def __init__(self, sources: xr.DataArray, geobox):
+        """Create a Tile representing a dataset that can be loaded.
+
+        :param xr.DataArray sources: An array of non-spatial dimensions of the request, holding lists of
+            datacube.storage.DatasetSource objects.
```

Comment:
Does this really expect a DataArray of DatasetSources?   Some of the code below seems to suggest that it is actually a DataArray of `datacube.model.Dataset` objects, e.g. as returned by `dc.group_datasets()`.

Reply:
Nope, the docs are wrong. Maybe at some point in the past...

Fixing.